src/init/network.rs

```diff
@@ -0,0 +1,281 @@
+ *   Unless required by applicable law or agreed to in writing, software      *
+ *   distributed under the License is distributed on an "AS IS" BASIS,        *
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ *   See the License for the specific language governing permissions and      *
+ *   limitations under the License.                                           *
+ *                                                                            *
+\* -------------------------------------------------------------------------- */
+
+use anyhow::anyhow;
+use futures::stream::TryStreamExt;
+use log::{error, info, warn};
+use std::collections::HashMap;
+use std::net::{Ipv4Addr, Ipv6Addr};
+use std::str;
+use std::{cmp, fs, io};
+
+use ipnetwork::IpNetwork;
+use ipnetwork::Ipv6Network;
+
+use rtnetlink::Handle;
+use netlink_packet_route::rtnl::link::nlas::Nla;
+
+#[derive(Eq, PartialEq)]
+enum IpType {
+    V6,
+    V4,
+}
+
+fn get_ip_type(ip: &Vec<u8>) -> Option<IpType> {
+    if ip.len() == 4 {
+        return Some(IpType::V4);
+    } else if ip.len() == 16 {
+        return Some(IpType::V6);
+    }
+    None
+}
+
+pub fn get_sriov_capabilities(iface: &str) -> Result<String, io::Error> {
+    fs::read_to_string(format!(
+        "/sys/class/net/{}/device/sriov_totalvfs",
+        iface
+    ))
+}
+
+async fn set_link_up(handle: Handle, iface: &str) -> Result<(), anyhow::Error> {
+    let mut links = handle.link().get().match_name(iface.to_string()).execute();
+
+    if let Some(link) = links.try_next().await? {
+        handle.link().set(link.header.index).up().execute().await?
+    } else {
+        warn!("iface '{}' not found", iface);
+    }
+    Ok(())
+}
+
+pub async fn add_address(
+    iface: &str,
+    ip: IpNetwork,
+    handle: rtnetlink::Handle,
+) -> Result<(), anyhow::Error> {
+    let mut links = handle.link().get().match_name(iface.to_string()).execute();
+
+    if let Some(link) = links.try_next().await? {
+        handle
+            .address()
+            .add(link.header.index, ip.ip(), ip.prefix())
+            .execute()
+            .await?
+    }
+
+    Ok(())
+}
+
+// Create max(limit, max possible sriov for given iface) sriov devices for the given iface
+pub fn setup_sriov(iface: &str, limit: u16) {
+    if limit == 0 {
+        return;
+    }
+
+    let sriov_totalvfs = match get_sriov_capabilities(iface){
+        Ok(val) => val,
+        Err(e) => {
+            error!("{}", e);
+            return
+        }  
+    };
+
+    let sriov_totalvfs = match sriov_totalvfs.trim_end().parse::<u16>(){
+        Ok(val) => val,
+        Err(e) => {
+            error!("{}", e);
+            return
+        }  
+    };
+
+    let num = cmp::min(limit, sriov_totalvfs);
+
+    fs::write(
+        format!("/sys/class/net/{}/device/sriov_numvfs", iface),
+        num.to_string(),
+    )
+    .expect("Unable to write file");
+}
+
+pub async fn init_iface(
```

Comment:
function signatures should denote the IP stack (IPv4) or (IPv6)

```rust
pub async fn init_iface_ipv6();
```

Reply:
I removed the init_iface functions completely now.
I think the initialization code of a given scenario can directly call the network functions.